### ISSUE TITLE
Trouble with compiling 

### ISSUE THREAD

**Issue by VelonacepsCalyxEggs** [SpaceManiac/hullrot#1]
Hello There! Basically I was trying to download this for my TG fork and test it with my friends, but I couldn't compile it. I'll say that I have almost zero experience with this stuff. I have Windows 10 (MSCV) as far as I know. I have all the stuff downloaded. 
This is the error it gave me while I was trying to build the cargo.toml.

```
PS C:\Users\User\.cargo> cargo build --release
warning: output filename collision.
The bin target `hullrot` in package `hullrot-server v0.2.0 (C:\Users\User\.cargo\src\hullrot-server)` has the same output filename as the lib target `hullrot` in package `hullrot-client v0.2.0 (C:\Users\User\.cargo\src\hullrot-client)`.
Colliding filename is: C:\Users\User\.cargo\target\release\deps\hullrot.pdb
The targets should have unique names.
Consider changing their names to be unique or compiling them separately.
This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
warning: output filename collision.
The bin target `hullrot` in package `hullrot-server v0.2.0 (C:\Users\User\.cargo\src\hullrot-server)` has the same output filename as the lib target `hullrot` in package `hullrot-client v0.2.0 (C:\Users\User\.cargo\src\hullrot-client)`.
Colliding filename is: C:\Users\User\.cargo\target\release\hullrot.pdb
The targets should have unique names.
Consider changing their names to be unique or compiling them separately.
This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
   Compiling openssl-sys v0.9.53
   Compiling opus-sys v0.2.1
   Compiling tempfile v3.1.0
   Compiling serde_derive v1.0.104
   Compiling mio v0.6.21
error: failed to run custom build command for `opus-sys v0.2.1`

Caused by:
  process didn't exit successfully: `C:\Users\User\.cargo\target\release\build\opus-sys-4be53358c0caac38\build-script-build` (exit code: 101)
  --- stderr
  thread 'main' panicked at 'program not found', C:\Users\User\.cargo\registry\src\github.com-1ecc6299db9ec823\opus-sys-0.2.1\build.rs:73:23
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
error: failed to run custom build command for `openssl-sys v0.9.53`

Caused by:
  process didn't exit successfully: `C:\Users\User\.cargo\target\release\build\openssl-sys-62fef51abf16687c\build-script-main` (exit code: 101)
  --- stdout
  cargo:rustc-cfg=const_fn
  cargo:rerun-if-env-changed=X86_64_PC_WINDOWS_MSVC_OPENSSL_LIB_DIR
  X86_64_PC_WINDOWS_MSVC_OPENSSL_LIB_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_LIB_DIR
  OPENSSL_LIB_DIR unset
  cargo:rerun-if-env-changed=X86_64_PC_WINDOWS_MSVC_OPENSSL_INCLUDE_DIR
  X86_64_PC_WINDOWS_MSVC_OPENSSL_INCLUDE_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_INCLUDE_DIR
  OPENSSL_INCLUDE_DIR unset
  cargo:rerun-if-env-changed=X86_64_PC_WINDOWS_MSVC_OPENSSL_DIR
  X86_64_PC_WINDOWS_MSVC_OPENSSL_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_DIR
  OPENSSL_DIR unset
  note: vcpkg did not find openssl as libcrypto and libssl: Aborted because VCPKGRS_DYNAMIC is not set
  note: vcpkg did not find openssl as ssleay32 and libeay32: Aborted because VCPKGRS_DYNAMIC is not set

  --- stderr
  thread 'main' panicked at '

  Could not find directory of OpenSSL installation, and this `-sys` crate cannot
  proceed without this knowledge. If OpenSSL is installed and this crate had
  trouble finding it,  you can set the `OPENSSL_DIR` environment variable for the
  compilation process.

  Make sure you also have the development packages of openssl installed.
  For example, `libssl-dev` on Ubuntu or `openssl-devel` on Fedora.

  If you're in a situation where you think the directory *should* be found
  automatically, please open a bug at https://github.com/sfackler/rust-openssl
  and include information about your system as well as this message.

  $HOST = x86_64-pc-windows-msvc
  $TARGET = x86_64-pc-windows-msvc
  openssl-sys = 0.9.53


  It looks like you're compiling for MSVC but we couldn't detect an OpenSSL
  installation. If there isn't one installed then you can try the rust-openssl
  README for more information about how to download precompiled binaries of
  OpenSSL:

  https://github.com/sfackler/rust-openssl#windows

  ', C:\Users\User\.cargo\registry\src\github.com-1ecc6299db9ec823\openssl-sys-0.9.53\build\find_normal.rs:150:5
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

**Comment by SpaceManiac** [comment]
For MSVC, your best bet is to follow these directions to satisfy the dependency: https://github.com/sfackler/rust-openssl/tree/openssl-sys-v0.9.36#windows-msvc

Newer versions of openssl-sys can install OpenSSL from vcpkg automatically but you would have to update Hullrot to that new version.

When I was developing the project I used the -gnu version of rustc with MSYS2, which has OpenSSL binary packages that I used.

If all else fails, I dug up old binaries from my archives and posted [v0.2.0](https://github.com/SpaceManiac/hullrot/releases/tag/v0.2.0) to the Releases page ([direct download link](https://github.com/SpaceManiac/hullrot/releases/download/v0.2.0/hullrot-win64-2020-10-19.zip)).

**Comment by VelonacepsCalyxEggs** [comment]
Understood, I'll try that as soon as I can.

**Comment by VelonacepsCalyxEggs** [comment]
> For MSVC, your best bet is to follow these directions to satisfy the dependency: https://github.com/sfackler/rust-openssl/tree/openssl-sys-v0.9.36#windows-msvc
> 
> Newer versions of openssl-sys can install OpenSSL from vcpkg automatically but you would have to update Hullrot to that new version.
> 
> When I was developing the project I used the -gnu version of rustc with MSYS2, which has OpenSSL binary packages that I used.
> 
> If all else fails, I dug up old binaries from my archives and posted [v0.2.0](https://github.com/SpaceManiac/hullrot/releases/tag/v0.2.0) to the Releases page ([direct download link](https://github.com/SpaceManiac/hullrot/releases/download/v0.2.0/hullrot-win64-2020-10-19.zip)).

I am not sure how to set the argument... I've read about environment arguments a little, and didn't understand where to put them. Setting it after the --release it gives me an error.

![image](https://user-images.githubusercontent.com/48419312/164701171-99efaffc-5e4f-42e8-a57f-fb1f1fa7987d.png)

**Comment by SpaceManiac** [comment]
In Powershell, it would be:
```pwsh
$Env:OPENSSL_DIR = "C:\OpenSSL-Win64"
cargo build --release
```

**Comment by VelonacepsCalyxEggs** [comment]
$Env:OPENSSL_DIR = "C:\Program Files (x86)\OpenSSL-Win32"

> ```powershell
> cargo build --release
> ```

Thank you! It did help with OpenSS, but it said that it failed custom compile for it. And Opus Sys still didn't compile for some reason. Have you got any ideas why?



```
PS C:\Users\User\.cargo> $Env:OPENSSL_DIR = "C:\Program Files (x86)\OpenSSL-Win32"
PS C:\Users\User\.cargo> cargo build --release
warning: output filename collision.
The bin target `hullrot` in package `hullrot-server v0.2.0 (C:\Users\User\.cargo\src\hullrot-server)` has the same output filename as the lib target `hullrot` in package `hullrot-client v0.2.0 (C:\Users\User\.cargo\src\hullrot-client)`.
Colliding filename is: C:\Users\User\.cargo\target\release\deps\hullrot.pdb
The targets should have unique names.
Consider changing their names to be unique or compiling them separately.
This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
warning: output filename collision.
The bin target `hullrot` in package `hullrot-server v0.2.0 (C:\Users\User\.cargo\src\hullrot-server)` has the same output filename as the lib target `hullrot` in package `hullrot-client v0.2.0 (C:\Users\User\.cargo\src\hullrot-client)`.
Colliding filename is: C:\Users\User\.cargo\target\release\hullrot.pdb
The targets should have unique names.
Consider changing their names to be unique or compiling them separately.
This may become a hard error in the future; see <https://github.com/rust-lang/cargo/issues/6313>.
   Compiling openssl-sys v0.9.53
   Compiling opus-sys v0.2.1
error: failed to run custom build command for `opus-sys v0.2.1`

Caused by:
  process didn't exit successfully: `C:\Users\User\.cargo\target\release\build\opus-sys-4be53358c0caac38\build-script-build` (exit code: 101)
  --- stderr
  thread 'main' panicked at 'program not found', C:\Users\User\.cargo\registry\src\github.com-1ecc6299db9ec823\opus-sys-0.2.1\build.rs:73:23
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
warning: build failed, waiting for other jobs to finish...
The following warnings were emitted during compilation:

warning: expando.c

error: failed to run custom build command for `openssl-sys v0.9.53`

Caused by:
  process didn't exit successfully: `C:\Users\User\.cargo\target\release\build\openssl-sys-62fef51abf16687c\build-script-main` (exit code: 101)
  --- stdout
  cargo:rustc-cfg=const_fn
  cargo:rerun-if-env-changed=X86_64_PC_WINDOWS_MSVC_OPENSSL_LIB_DIR
  X86_64_PC_WINDOWS_MSVC_OPENSSL_LIB_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_LIB_DIR
  OPENSSL_LIB_DIR unset
  cargo:rerun-if-env-changed=X86_64_PC_WINDOWS_MSVC_OPENSSL_INCLUDE_DIR
  X86_64_PC_WINDOWS_MSVC_OPENSSL_INCLUDE_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_INCLUDE_DIR
  OPENSSL_INCLUDE_DIR unset
  cargo:rerun-if-env-changed=X86_64_PC_WINDOWS_MSVC_OPENSSL_DIR
  X86_64_PC_WINDOWS_MSVC_OPENSSL_DIR unset
  cargo:rerun-if-env-changed=OPENSSL_DIR
  OPENSSL_DIR = C:\Program Files (x86)\OpenSSL-Win32
  cargo:rustc-link-search=native=C:\Program Files (x86)\OpenSSL-Win32\lib
  cargo:include=C:\Program Files (x86)\OpenSSL-Win32\include
  OPT_LEVEL = Some("3")
  TARGET = Some("x86_64-pc-windows-msvc")
  HOST = Some("x86_64-pc-windows-msvc")
  CC_x86_64-pc-windows-msvc = None
  CC_x86_64_pc_windows_msvc = None
  HOST_CC = None
  CC = None
  CFLAGS_x86_64-pc-windows-msvc = None
  CFLAGS_x86_64_pc_windows_msvc = None
  HOST_CFLAGS = None
  CFLAGS = None
  CRATE_CC_NO_DEFAULTS = None
  CARGO_CFG_TARGET_FEATURE = Some("fxsr,llvm14-builtins-abi,sse,sse2")
  DEBUG = Some("false")
  running: "C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\VC\\Tools\\MSVC\\14.30.30705\\bin\\HostX64\\x64\\cl.exe" "-nologo" "-MD" "-O2" "-Brepro" "-I" "C:\\Program Files (x86)\\OpenSSL-Win32\\include" "-W4" "-E" "build/expando.c"
  cargo:warning=expando.c
  exit code: 0

  --- stderr
  thread 'main' panicked at 'assertion failed: version.starts_with(\"0x\")', C:\Users\User\.cargo\registry\src\github.com-1ecc6299db9ec823\openssl-sys-0.9.53\build/main.rs:258:5
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
PS C:\Users\User\.cargo>
```

**Comment by SpaceManiac** [comment]
For Opus, it looks like you ran in to https://github.com/SpaceManiac/opus-rs/issues/16. I've pushed a fix for it.

I don't really know what's going on with OpenSSL there. I tried out openssl-sys's vcpkg support but it didn't work out of the box (vcpkg installs `x86-windows` variant but the crate wants `x86-windows-static-md`??) and the documentation is underwhelming. You'll have to experiment yourself I think.

BTW, in case it comes up later, make sure you build `hullrot-client` in 32-bit mode or it won't be loadable by BYOND. The server works either way.
